### ISSUE TITLE
Occasional crashes in UserLocationOperation

### ISSUE THREAD

**Issue by jshier** [danthorpe/Operations#151]
Although not as regular as #121, my project is currently seeing crashes in `UserLocationOperation`'s `didUpdateLocations` handler.

Latest crash looks like this (sorry, Crashlytics isn't very good for Swift code):
```
0 libobjc.A.dylib objc_retain + 16
1 Operations _TFC10Operations22_UserLocationOperation15locationManageruRq_S_19LocationManagerType_fGS0_q__FTCSo17CLLocationManager18didUpdateLocationsGSaCSo10CLLocation__T_ + 660
2 Operations _TToFC10Operations22_UserLocationOperation15locationManageruRq_S_19LocationManagerType_fGS0_q__FTCSo17CLLocationManager18didUpdateLocationsGSaCSo10CLLocation__T_ + 120
```
Another looks like this:
```
0 libobjc.A.dylib objc_retain + 16
1 Operations _TPA__TFFC10Operations22_UserLocationOperation15locationManageruRq_S_19LocationManagerType_FGS0_q__FTCSo17CLLocationManager18didUpdateLocationsGSaCSo10CLLocation__T_U_FT_T_ + 40
2 libdispatch.dylib _dispatch_call_block_and_release + 24
```
Seems like there's either a memory issue with the `CLLocationManager` instance used by the operation, or something in the way I'm using it that is capturing or not capturing things like it should. Any guidance? I have yet to be able to replicate the crash while in the debugger.

**Comment by danthorpe** [comment]
Sorry about that @jshier - what version is this?

**Comment by danthorpe** [comment]
Also, if you can share how you're using it, that would also be great. E.g. are you using the completion block, or the property to access the `CLLocation`?

**Comment by jshier** [comment]
This is version 2.5, with the latest version of the `LocationManager` I showed in the previous issue:
```swift
import Operations

typealias LocationCompletionHandler = (location: CLLocation) -> Void

class LocationManager : OperationQueue {
    private static let sharedManager = LocationManager()
    
    static var lastUserLocation: CLLocation?
    
    static func updateUserLocation(accuracy: CLLocationAccuracy = kCLLocationAccuracyThreeKilometers) {
        currentUserLocation(accuracy) { _ in }
    }
    
    static func currentUserLocation(accuracy: CLLocationAccuracy = kCLLocationAccuracyThreeKilometers, completionHandler: LocationCompletionHandler) {
        let locationOperation = UserLocationOperation(accuracy: accuracy) { location in
            lastUserLocation = location
            completionHandler(location: location)
        }
        sharedManager.addOperation(locationOperation)
    }
}
```
Used like this:
```swift
LocationManager.updateUserLocation()
```
or
```swift
guard let lastKnownCoordinates = LocationManager.lastUserLocation?.coordinate else {
    return nil
}
```

**Comment by danthorpe** [comment]
Okay @jshier will take a look at this. I take you have not been able to reproduce this, and are just going via crash logs from Crashlytics?

**Comment by jshier** [comment]
I haven't, no, so we just have logs from Crashlytics. However, this app isn't public yet, so we've only been running into this issue during QA. Even a low number crashes is rather concerning with so little overall usage.

**Comment by danthorpe** [comment]
Hi @jshier I know this is a bit of a pain - but is there any chance you can run a build using branch `bug/OPR-151_crash_in_location_manager`?

I've made the following changes:

1. When the operation stops updating locations (called just before it finishes) it also sets the location manager's delegate to `nil`.

2. Using weak references in all places that capture `self` - which where it dispatch to the main queue. 

Also, regarding Crashlytics - are you using a custom logger with Operations which pipe the logs though to Crashlytics? If so, can you set the log severity on the `UserLocationOperation` inside your manager to `.Verbose` so that we can see where in the lifecycle of the operation the crash is occurring?

if you are not using a custom logger - see: [custom logging](https://operations.readme.io/docs/custom-logging) in the programming guide, and this page on [`CLS_Log`](http://support.crashlytics.com/knowledgebase/articles/92519-how-do-i-use-logging).

If you cannot create a build, then I might just make a start on release 2.5.1 to include this.

**Comment by jshier** [comment]
We're testing the fix branch now. I'll take a look into the logging as well.

**Comment by jshier** [comment]
We were able to see this fairly regularly in the simulator, and the fix branch seems to have fixed it entirely. I think it's more to do with setting the delegate to nil than the weak capture, but whatever it was, it seems fixed now. Thanks.

**Comment by danthorpe** [comment]
Okay, great. Will merge and release 2.5.1 in the morning. Thanks for reporting & testing!

**Comment by danthorpe** [comment]
Closing this as it's now fixed in the 2.5.1 release.
